File: src/local_auth.zig
//! LocalAuthentication bridge — biometric (Touch ID) prompt for unlock and
//! `get_secret`. The actual ObjC code lives in `local_auth.m`; this file is
//! a thin extern wrapper plus a Zig-friendly API.
//!
//! Why ObjC: the LAContext API uses Objective-C blocks for its async reply,
//! and dispatch_semaphore to pin it down to a synchronous call. Hand-rolling
//! both via objc_msgSend in Zig is significantly more code and several
//! footguns (Block_layout struct, _NSConcreteGlobalBlock, NSString lifetime)
//! than just letting the system Objective-C compiler handle it.

const std = @import("std");

extern "c" fn secretctl_la_available() c_int;
extern "c" fn secretctl_la_evaluate(reason: [*:0]const u8) c_int;

/// True if Touch ID / Face ID is available and enrolled on this device.
pub fn available() bool {
    return secretctl_la_available() == 1;
}

/// Synchronously prompt for biometric authentication. Blocks until the user
/// responds (success / cancel) or the OS times out. Returns true on success.
/// `reason` is shown in the system prompt; passed as a NUL-terminated string.
pub fn evaluate(reason: [*:0]const u8) bool {
    return secretctl_la_evaluate(reason) == 1;
}

test "available is callable" {
    // We don't assert true/false because CI runners may or may not have
    // Touch ID hardware; just call into it to make sure the FFI links.
    _ = available();
}

File: src/local_auth.m
// LocalAuthentication.framework bridge.
//
// Implementing this in Objective-C rather than Zig avoids hand-rolling
// objc_msgSend, NSString construction, and the Block ABI.
// Two C-ABI exports are sufficient for our use cases (Touch ID-gated
// keychain unlock and MCP `get_secret`).
//
// Build flag in build.zig: -fobjc-arc + -framework LocalAuthentication.

#import <LocalAuthentication/LocalAuthentication.h>
#include <dispatch/dispatch.h>

// Returns 1 if biometrics (Touch ID / Face ID) are available and enrolled,
// else 0.
int secretctl_la_available(void) {
    LAContext *ctx = [[LAContext alloc] init];
    NSError *err = nil;
    BOOL ok = [ctx canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
                               error:&err];
    return ok ? 1 : 0;
}

// Synchronously evaluate biometric policy.
// Returns:
//   1  on user-authenticated success
//   0  on cancel / failure / not-available
//
// `reason` is the localized message shown in the system prompt.
int secretctl_la_evaluate(const char *reason) {
    LAContext *ctx = [[LAContext alloc] init];
    NSError *err = nil;
    if (![ctx canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
                          error:&err]) {
        return 0;
    }

    NSString *r = [NSString stringWithUTF8String:reason];
    __block BOOL result = NO;
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);

    [ctx evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
        localizedReason:r
                  reply:^(BOOL success, NSError * _Nullable error) {
                      (void)error;
                      result = success;
                      dispatch_semaphore_signal(sem);
                  }];

    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
    return result ? 1 : 0;
}
